### PR TITLE
Fix TypeScript project builds with Generated Types

### LINKED ISSUE #17084: types autogeneration fails on adding new content types
### Required System information

- Node.js version: 18.16.0
- NPM version: 9.5.1
- Strapi version: 4.11.2
- Database: postgres 
- Operating system: windows
- Is your project Javascript or Typescript: typescript

### Describe the bug

Related to the new generating type system: https://github.com/strapi/strapi/pull/16917
Enabling this feature, autogeneration seems to be not correctly handled.
Running strapi in development mode and adding a new content type, the server reloads but it crashes because of new collection types are not added to contentTypes.d.ts .

### Steps to reproduce the behavior
0. add config/typescript.ts file with the following content:
`
export default ({ env }) => ({
  autogenerate: true,
});
`
1. generate types first time using "npm run strapi ts:generate-types"
2. run strapi using "npm run develop"
3. Go to admin/plugins/content-type-builder/content-types
4. Add a new collection with any name and any field(s).
5. Save, the server will try to restart
6. See error:
Argument of type '"api::entityname.entityname"' is not assignable to parameter of type 'ContentType'.
 export default factories.createCoreService('api::entityname.entityname');

### Expected behavior

Strapi should generate the new content types before restarting the server, avoiding the typescript error.

### Screenshots

image
image

### LINKED ISSUE #17906: Potential regression introduced by #17799
### Required System information

- Node.js version: `v18.17.0`
- NPM version: `9.6.7`
- Strapi version: `4.13.1`
- Database: `SQLite`
- Operating system: `macOS 13.4.1`
- Is your project Javascript or Typescript: Typescript

### Describe the bug

New types introduced by #17799 mark `findMany(...)`'s return type to `Promise<AnyEntity | AnyEntity[]>` which sounds weird since it should always return an array.

### Steps to reproduce the behavior

The picture below shows a very simple controller that I created in a test project. I fetch the `api::post.post` objects and then I loop through them with both `.map(...)` and a for loop to show the issue. No TS types have been generated  in this case.

**Picture A:**
Screenshot 2023-09-02 at 02 09 22

The problem can be solved by generating the types using `strapi ts:generate-types` as you can see in the picture below...

**Picture B:**
Screenshot 2023-09-02 at 02 12 09

But I had to comment out the for loop before running `strapi ts:generate-types` otherwise it fails due to a typecheck error:
```
Type 'AnyEntity | AnyEntity[]' must have a '[Symbol.iterator]()' method that returns an iterator.ts(2488)
```

Commenting out pieces of code any time you wanna build the types is impossible in a complex project...

### Expected behavior

I would expect the following behavior:
1. `findMany(...)` should always return an array (this might apply to other entity service functions too, but I didn't investigate further)
2. currently, the `strapi ts:generate-types` command needs to compile the whole codebase before the types can be generated, I find this pretty inconvenient because it makes the compilation fail due to the lack of types, but the types generation fails too because it can't compile the code... I face the same issue any time I create a new content type from the admin dashboard while having the types already built from the previous codebase: the server needs to restart after the code is compiled again, but the old types do not include the new content type so the compilation fails and the server dies. The only solution is to remove the types folder and re-generate the types

I hope it's just me doing something wrong here, but I couldn't find a proper solution...

Thank you for your time.

---

## Edit (04-09-2023)

I also noticed that a the following pattern stopped working with the new version:
Screenshot 2023-09-04 at 19 22 52

Using the `id` field here is pretty useful, especially when other fields are optional.

### LINKED ISSUE #17962: @strapi/strapi/lib/index.d.ts missing compile() function
## Bug report

image

image

### Required System information

- Node.js version:
- NPM version:
- Strapi version: 4.12.7
- Database:
- Operating system:
- Is your project Javascript or Typescript: Typescript

### Describe the bug

As you can see, the index.js mounted a compile() function to Strapi, but the index.d.ts didn't. 

### Steps to reproduce the behavior

1. Checkout the code
2. Open lib/index.d.ts and lib/index.js

### Expected behavior

index.d.ts should sync with index.js.

### PR DESCRIPTION
**Pull request by Convly**
### What does it do?

- [DEVELOP MODE] Prevent typescript apps from crashing upon start or re-start when generated types are not up-to-date with the current context 
- [CLI] Allow generating types from the CLI even if the applications have compilation errors
- [NEW] Always generate application types on dev mode

### Why is it needed?

- Apps with generated types not matching the current state of their codebase cannot re-generate types because it currently asks for a successful compilation. The same thing happens on app restart (e.g. after creating a CT from the CTB) where newly created controllers & services don't match existing types (not generated yet)

### How to test it?

- Start an app in dev mode, and create a content type. The server should restart successfully & update the generated types accordingly
- Remove any reference to the newly created content type from the generated types, run `yarn strapi ts:generate-types`. It should throw an error and instead sync the types again

Edge cases:
- The app should still fail to start if there is another compilation error (not related to generated types)
- The app should successfully update types & start even if it's out of sync at startup (still dev mode only)

### Related issue(s)/PR(s)

fix #17084
fix #17906
fix #17962

### CODE REVIEW

File: packages/utils/typescript/lib/compilers/index.js [hunk 1]
```diff
@@ -1,9 +1,7 @@
 'use strict';
 
 const basic = require('./basic');
-const watch = require('./watch');
```
Comment (innerdvations) [inline packages/utils/typescript/lib/compilers/index.js:4]: Are we sure we need to remove watch for this to work?

**Comment by Convly** [comment]
Experimental release to test the approach: `0.0.0-experimental.3cadca87513eebfb388e8328d166fc3112020c51`

**Comment by liarco** [comment]
Hi @Convly, thank you for your time working on this. I just checked the experimental version on a local reproducer and it worked fine.

Some details about the test:
- small project with a couple of entities and dummy content
- custom controllers using types and requiring TS types to compile successfully
- tested `yarn build` with no TS types -> build failed due to the missing types (as expected)
- tested `yarn strapi ts:generate-types && yarn build` -> both types generation and the build worked fine (as expected)

Without these changes the last command fails at types generation.

I hope this helps.

**Comment by Convly** [comment]
> Hi @Convly, thank you for your time working on this. I just checked the experimental version on a local reproducer and it worked fine.

Awesome. we'll probably do some more QA with the team on various other scenarios but I'm glad you could confirm it fixes your issue.

> I hope this helps.

It does, thanks a lot!

**Comment by innerdvations** [comment]
Just added types for strapi.compile

**Review by innerdvations — APPROVED** [review]
LGTM! Let's get some more QA on it before merging.

**Pull request merged**